File: leetcode/arrays_and_hashing/golang/49_group_anagrams.go
/*
Given an array of strings strs, group the anagrams together. You can return the answer in any order.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.



Example 1:

Input: strs = ["eat","tea","tan","ate","nat","bat"]
Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
*/

// This solution uses a map to store the anagrams. The key is the sorted string and the value is the list of anagrams.
// Time complexity: O(n * klogk) where n is the length of strs and k is the length of the longest string in strs
func groupAnagrams(strs []string) [][]string {
	m := make(map[string][]string)
	for _, str := range strs {
		key := getKey(str)
		m[key] = append(m[key], str)
	}

	res := make([][]string, 0)
	for _, v := range m {
		res = append(res, v)
	}
	return res
}

// Optimized solution
// Time complexity: O(n * k) where n is the length of strs and k is the length of the longest string in strs
func groupAnagrams(strs []string) [][]string {
	// Create a map to store the anagrams
	// The key is an array of 26 integers representing the count of each character in the string
	// The value is the list of anagrams
	m := make(map[[26]int][]string)
	for _, str := range strs {
		// Get the key for the string
		key := getKey(str)
		// Append the string to the list of anagrams
		m[key] = append(m[key], str)
	}

	// Convert map to slice
	res := make([][]string, 0)
	for _, v := range m {
		// Append the list of anagrams to the result
		res = append(res, v)
	}
	return res
}

File: leetcode/arrays_and_hashing/python/arrays_and_hashing/49_group_anagrams.py
# Given an array of strings strs, group the anagrams together. You can return the answer in any order.
# 
# An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.
# 
# Example 1:
# 
# Input: strs = ["eat","tea","tan","ate","nat","bat"]
# Output: [["bat"],["nat","tan"],["ate","eat","tea"]]


# My solution by sorting each string and using that as the key in a dictionary.
# Time complexity: O(n*mlogm) where n is the number of strings and m is the length of the longest string.

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        ans = {}
        for string in strs:
            s_str = ''.join(sorted(string))
            if s_str in ans:
                ans[s_str].append(string)
            else:
                ans[s_str] = [string]
        return ans.values()

# Time optimized solution using count array and ord function to get the frequency of each character in the string.
# Time complexity: O(n*m) where n is the number of strings and m is the length of the longest string.
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        ans = {}
        for string in strs:
            count = [0]*26
            for c in string:
                # Subtracting the ASCII value of 'a' to get the index of the character in the count array.
                count[ord(c) - ord('a')] += 1
            s_str = tuple(count)
            if s_str in ans:
                ans[s_str].append(string)
            else:
                ans[s_str] = [string]
        return ans.values()